### PR TITLE
don't lint on all caps enum variants starting with the same char

### LINKED ISSUE #665: Example of too demanding 1-letter `enum_variant_names` warning
```
src/evm/tests.rs:28:1: 30:2 warning: All variants have the same prefix: `C`, #[warn(enum_variant_names)] on by default
src/evm/tests.rs:28 enum FakeCallType {
src/evm/tests.rs:29     CALL, CREATE
src/evm/tests.rs:30 }
```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #665, #677

### CODE REVIEW

File: src/enum_variants.rs [hunk 1]
```diff
@@ -40,40 +41,39 @@ fn partial_rmatch(left: &str, right: &str) -> usize {
...
+    #[allow(while_let_on_iterator)]
     fn check_item(&mut self, cx: &EarlyContext, item: &Item) {
         if let ItemKind::Enum(ref def, _) = item.node {
             if def.variants.len() < 2 {
                 return;
             }
             let first = var2str(&def.variants[0]);
-            let mut pre = first.to_string();
-            let mut post = pre.clone();
+            let mut pre = first[..camel_case_until(&*first)].to_owned();
```
Comment (Manishearth) [inline src/enum_variants.rs]: Why is `.to_owned()` necessary here?

File: src/utils/mod.rs [hunk 2]
```diff
@@ -607,3 +607,65 @@ pub fn is_expn_of(cx: &LateContext, mut span: Span, name: &str) -> Option<Span>
         }
     }
 }
+
+// returns index until which the str is camel case
```
Comment (Manishearth) [inline src/utils/mod.rs]: Perhaps "returns index after first CamelCase component of string"?

File: src/utils/mod.rs [hunk 3]
```diff
@@ -607,3 +607,65 @@ pub fn is_expn_of(cx: &LateContext, mut span: Span, name: &str) -> Option<Span>
...
+        }
+    }
+    if up {
+        last_i
+    } else {
+        s.len()
+    }
+}
+
+// returns index from which on the str is camel case
```
Comment (Manishearth) [inline src/utils/mod.rs]: ditto

**Comment by Manishearth** [comment]
r=me with the comment fix

**Comment by oli-obk** [comment]
needs a full review, since it now also addresses #677

File: src/utils/mod.rs [hunk 4]
```diff
@@ -607,3 +607,65 @@ pub fn is_expn_of(cx: &LateContext, mut span: Span, name: &str) -> Option<Span>
         }
     }
 }
+
+// returns index after first CamelCase component of string
```
Comment (mcarton) [inline src/utils/mod.rs]: Small pedantic nit: Should be a *rustdoc* `///` comment and the convention seems to be full sentences with a capital letter and dot.

Comment (oli-obk) [inline src/utils/mod.rs]: indeed. Can we lint on comments before functions/types/modules that should be doc-comments? Comments don't show up in the ast afaik, so this is probably messy :(

File: src/enum_variants.rs [hunk 5]
```diff
@@ -31,48 +32,63 @@ fn var2str(var: &Variant) -> InternedString {
...
 }
 
-fn partial_rmatch(left: &str, right: &str) -> usize {
-    left.chars().rev().zip(right.chars().rev()).take_while(|&(l, r)| l == r).count()
+fn partial_rmatch(post: &str, name: &str) -> usize {
+    // skip(1) to ensure that the postfix never takes the whole variant name
+    post.chars().rev().zip(name.chars().skip(1).rev()).take_while(|&(l, r)| l == r).count()
+}*/
+
+fn partial_match(pre: &str, name: &str) -> usize {
```
Comment (Manishearth) [inline src/enum_variants.rs]: We actually could fold the camel_case methods into this. All we need to do is ensure that the next character is uppercase and at least one of the current ones is not.

**Pull request merged**